macaroon-thought.md

```diff
@@ -0,0 +1,399 @@
+
+1. Start with an `(org=4721, mask=*)` token, and attenuate with 
+   `(app=555, mask=*)` to make an app-specific token; give that token
+   to the developers of that particular app.
+   
+2. Start with an `(org=4721, mask=*)` token, and attenuate with 
+   `(org=4721, mask=r)` to make a read-only token; give that token
+   to an auditor (you can get more specific here, this is just to get
+   you thinking).
+   
+3. Start with `(org=4721, mask=*) (app=555, mask=*)` and attenuate
+   with `(validity-window 2 hours)` to give time-limited access to
+   an application to a ops team member investigating an outage.
+   
+## Entry-Level Macaroon Cryptography
+
+Assume some HMAC secret `R` held by our API. You can see right away how our API could HMAC a single caveat, like `(org=4721, mask=*)`, or even a bunch of them, producing an HMAC tag `T`, tacked onto the message. Even Rails can do that. You as a user can't change these caveats or add to them without breaking `T`. We want to do better.
+
+Instead: start by using `R` to HMAC a nonce, to produce `T_0`. Tack `T_0` on to the end of the message, like so:
+
+```
+    Nonce
+    T_0
+```
+
+You can see how the API could verify `T_0` and thus the nonce.
+
+Now, add the caveat `(org=4721, mask=*)`: take `T_0` off the bottom of the message and use it as the key, instead of `R`, to HMAC the caveat, producing `T_1`. We end up with this message:
+
+```
+    Nonce
+    (org=4721, mask=*)
+    T_1
+```
+
+Notice that `T_0` is no longer in the message. Because you don't know `T_0` anymore, you can't reproduce the inputs to `T_1`; the `(org=4721, mask=*)` is stuck there permanently now.
+
+Our API doesn't have this problem, because it knows `R`, the original secret used to sign the nonce. When it gets an attenuated Macaroon, it can simply HMAC the Nonce with `R` to get `T_0`, and then use that to HMAC the org caveat to get `T_1`, and check if it matches the tag on the message.
+
+There you go: a token our API can verify, and that you as a user can extend with additional caveats, by taking the tag off the end of the Macaroon, using it as a key to HMAC your new caveat, and putting that new HMAC at the end of the message. Pretty slick.
+
+## IfPresent Caveats
+
+Another obvious thing you'd want to attenuate a token to: a token that
+can only be used to deploy applications, and nothing else, so you can
+hand that token off to your CI/CD system.
+
+This should be easily expressed in caveats, except that the access you
+need to deploy an application is quirky: Fly.io deployments involve
+_builders_ (temporary Fly Machines that run Docker container builders) and _wireguard access_ (to talk to the builder). 
+
+We have caveats for both of these things:
+
+```
+  ((feature-set "builders", mask=*), (feature-set "wg", mask=*))
+```
+
+So far so good. Here's the problem, though, when we try to assemble this into a single token:
+
+```
+   (org=4721, mask=*)
+   ((feature-set "builders", mask=*), (feature-set "wg", mask=*))
```

Comment:
I wonder if it would be worth using a shorthand that is slightly closer to our actual code. 

```golang
FeatureSet{"builders": PermissionAll, "wg": PermissionAll}
```

is pretty close to what you'd actually write and might be clearer than the `((), ())` notation for describing a single caveat referring to multiple things. Even for organization caveats, it's not that much more verbose

```golang
Organization{4721: PermissionAll}
```